exporter/signalfxexporter/exporter.go

```diff
@@ -0,0 +1,213 @@
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package signalfxexporter
+
+import (
+	"bytes"
+	"compress/gzip"
+	"context"
+	"errors"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"net/http"
+	"sync"
+	"time"
+
+	"github.com/golang/protobuf/proto"
+	"github.com/open-telemetry/opentelemetry-collector/consumer/consumerdata"
+	"github.com/open-telemetry/opentelemetry-collector/consumer/consumererror"
+	"github.com/open-telemetry/opentelemetry-collector/exporter"
+	"github.com/open-telemetry/opentelemetry-collector/exporter/exporterhelper"
+	sfxpb "github.com/signalfx/com_signalfx_metrics_protobuf"
+	"go.uber.org/zap"
+)
+
+// New returns a new SignalFx exporter.
+func New(
+	config *Config,
+	logger *zap.Logger,
+) (exporter.MetricsExporter, error) {
+
+	if config == nil {
+		return nil, errors.New("nil config")
+	}
+
+	if config.Name() == "" {
+		config.SetType(typeStr)
+		config.SetName(typeStr)
+	}
+
+	if config.Realm == "" {
+		err := fmt.Errorf(
+			"%q config requires a non-empty \"realm\"",
+			config.Name())
+		return nil, err
+	}
+
+	if config.Timeout == 0 {
+		config.Timeout = 5 * time.Second
+	}
+
+	if config.Timeout < 0 {
+		err := fmt.Errorf(
+			"%q config cannot have a negative \"timeout\"",
+			config.Name())
+		return nil, err
+	}
+
+	headers, err := buildHeaders(config)
+	if err != nil {
+		return nil, err
+	}
+
+	url := fmt.Sprintf("https://ingest.%s.signalfx.com/v2/datapoint", config.Realm)
+
+	s := &httpSender{
+		url:     url,
+		headers: headers,
+		client: &http.Client{
+			// TODO: What other settings of http.Client to expose via config?
+			//  Or what others change from default values?
+			Timeout: config.Timeout,
+		},
+		logger: logger,
+		zippers: sync.Pool{New: func() interface{} {
+			return gzip.NewWriter(nil)
+		}},
+	}
+
+	exp, err := exporterhelper.NewMetricsExporter(
+		&config.ExporterSettings,
+		s.pushMetricsData,
+		exporterhelper.WithTracing(true),
+		exporterhelper.WithMetrics(true))
+
+	return exp, err
+}
+
+// httpSender sends the data to the SignalFx backend.
+type httpSender struct {
+	url     string
+	headers map[string]string
+	client  *http.Client
+	logger  *zap.Logger
+	zippers sync.Pool
+}
+
+func (s *httpSender) pushMetricsData(
+	ctx context.Context,
+	md consumerdata.MetricsData,
+) (droppedTimeSeries int, err error) {
+
+	sfxDataPoints, numDroppedTimeseries, err := metricDataToSingalFxV2(s.logger, md)
+	if err != nil {
+		return exporterhelper.NumTimeSeries(md), consumererror.Permanent(err)
+	}
+
+	body, compressed, err := s.encodeBody(sfxDataPoints)
+	if err != nil {
+		return exporterhelper.NumTimeSeries(md), consumererror.Permanent(err)
+	}
+
+	req, err := http.NewRequest("POST", s.url, body)
+	if err != nil {
+		return exporterhelper.NumTimeSeries(md), consumererror.Permanent(err)
+	}
+
+	if s.headers != nil {
```

Comment:
`range nil` is a safe noop so this guard is unnecessary. golangci-lint will flag it as a violation as well.